app_dart/lib/src/service/scheduler.dart

```diff
@@ -154,6 +154,10 @@ class Scheduler {
       log.severe('Failed to add commit ${commit.sha!}: $error');
     }
 
+    await luciBuildService.schedulePostsubmitBuilds(commit: commit, toBeScheduled: toBeScheduled);
+    // Update tasks for those that were scheduled
+    await datastore.insert(tasks);
```

Comment:
why do we need to insert again?

Reply:
LuciBuildService updates the task statuses once triggered

Reply:
I may miss it, but which section updates the task status?

Reply:
My bad, it's in https://cs.opensource.google/flutter/cocoon/+/main:app_dart/lib/src/service/scheduler.dart;l=140

The logic is:
1. Insert commit and initial targets to datastore to generate keys
2. Trigger the builds based on priority
3. Update datastore again with the final status (some builds will be scheduled due to batch scheduling)

Reply:
but `tasks` are already updated at line 149?

Reply:
This PR moves it around. In `schedulePostsubmitBuilds` these entities need to be written beforehand to get their keys (since they're deterministic, datastore generates them). See https://github.com/flutter/flutter/issues/78876

Reply:
I think the confusion here ` await luciBuildService.schedulePostsubmitBuilds(commit: commit, toBeScheduled: toBeScheduled);`.

As I see the logic, it only pushes requests to pub sub where I don't see any task status update.

Reply:
Correct. We preemptively need to mark these tasks as in progress as there's no gurantees when LUCI will send a notification acking them as scheduled. This is to ensures batching is maintained

Reply:
To wrap up what I am talking about:

https://github.com/flutter/cocoon/blob/66667f15396ee7d6fd46d27bc47dd62752f3f93e/app_dart/lib/src/service/scheduler.dart#L133-L159

line140: task status changes
line149: tasks updated in datastore
line157: no task status change
line159: why do we need to update tasks in datastore again?

Reply:
Thanks for your patience on this, Friday brain. I removed the extra write.